### ISSUE TITLE
[AVR] integer division incorrectly yields result value 1

### ISSUE THREAD

**Issue by crclark96** [rust-lang/rust#82242]
I'm writing some code for arduino using the avr-hal crate, and when I print the result of a division operation, I'm getting the result value of 1 for almost every combination of numbers. Exceptions that I've found so far include when the divisor is a power of two, which I'm assuming is being optimized to a bitshift operation (this is not happening in my minimum code example, I do not know why but it happened in another context so I'm including this detail here in case it comes up again).

I tried this code:

```rust
#![no_std]
#![no_main]

extern crate panic_halt;

use arduino_uno::prelude::*;

#[arduino_uno::entry]
fn main() -> ! {
    let dp = arduino_uno::Peripherals::take().unwrap();

    let mut pins = arduino_uno::Pins::new(
        dp.PORTB,
        dp.PORTC,
        dp.PORTD,
    );
    let mut serial = arduino_uno::Serial::new(
        dp.USART0,
        pins.d0,
        pins.d1.into_output(&mut pins.ddr),
        57600.into_baudrate(),
    );

    for x in (5..100).step_by(5) {
        for y in 1..5 {
            let z = x / y;
            ufmt::uwriteln!(&mut serial, "{} / {} = {}\r", x, y, z).void_unwrap();
        }
    }
    loop {}
}
```
Output:
```
5 / 1 = 1
5 / 2 = 1
5 / 3 = 1
5 / 4 = 1
10 / 1 = 1
10 / 2 = 1
10 / 3 = 1
10 / 4 = 1
-- snip --
90 / 1 = 1
90 / 2 = 1
90 / 3 = 1
90 / 4 = 1
95 / 1 = 1
95 / 2 = 1
95 / 3 = 1
95 / 4 = 1
```

### Meta

`rustc --version --verbose`:
```
rustc 1.51.0-nightly (c2de47a9a 2021-01-06)
binary: rustc
commit-hash: c2de47a9aa4c9812884f341f1852e9c9610f5f7a
commit-date: 2021-01-06
host: x86_64-unknown-linux-gnu
release: 1.51.0-nightly
```

I'm using nightly-2021-01-07 since AVR support has been broken since then. The code has to be built with the `--release` flag due to this issue: [https://github.com/rust-lang/compiler-builtins/issues/400](https://github.com/rust-lang/compiler-builtins/issues/400)

Here's the repo with the code and configs for AVR: [https://github.com/crclark96/avr-division-repro](https://github.com/crclark96/avr-division-repro). I'm using an Arduino Uno board for this.

**Comment by kolbma** [comment]
Any idea what might be responsible for this?  
I've tried 
```
nightly-2020-10-25-x86_64-unknown-linux-gnu
nightly-2020-11-25-x86_64-unknown-linux-gnu
nightly-2020-12-07-x86_64-unknown-linux-gnu
nightly-2020-12-12-x86_64-unknown-linux-gnu
nightly-2021-01-07-x86_64-unknown-linux-gnu
```
and division doesn't work.  
Could be possible that I have not used divisions in my rust-avr code before, but I can't believe that nobody has used it?  
Is it in the **_compiler_builtins_** crate?

Wondering why such nasty bugs are not caught in the compiler build pipeline...

**Comment by jacobmischka** [comment]
The example above does reproduce for me, I receive all 1s, but I also have been successful performing integer division in projects. [The math here for driving a speaker](https://github.com/jacobmischka/tea-whistle/blob/main/src/tone.rs), for example, does properly produce the correct tones on the same device I tested the above code on. I do not understand how both of those could be true, but it seems to be the case.

**Comment by kolbma** [comment]
Strange, here u8 and u16 is fine. But u32 definitively not. I've not checked the signed variants.  
It looks like memory corruption.  
I've got also crashes when writing the result of the division to the usart with hardcoded u32, with u16 there is no problem.

**Comment by jacobmischka** [comment]
Oh, I didn't include a `u32`, oops, I meant to. Yeah, `u32` uDisplay seems broken, but `to_le/be/ne_bytes()` works around that sort of.

**Comment by crclark96** [comment]
Just a heads up I filed this issue for the `ufmt` crate when printing certain numbers, so using `to_le/be/ne_bytes()` seems to be the most reliable for now

https://github.com/japaric/ufmt/issues/28

**Comment by jacobmischka** [comment]
> Strange, here u8 and u16 is fine. But u32 definitively not. I've not checked the signed variants.
> It looks like memory corruption.
> I've got also crashes when writing the result of the division to the usart with hardcoded u32, with u16 there is no problem.

You're right, they don't work with `u32` inputs. The reason why I was able to get it to work in my project I linked above is because they actually were `u16`s, just casted to `u32`.

```rust
    for x in (5u32..100u32).step_by(5) {
        for y in 1u32..5u32 {
            let z: u32 = x / y;
            uwriteln!(
                &mut serial,
                "{:?} / {:?} = {:?}\r",
                x.to_le_bytes(),
                y.to_le_bytes(),
                z.to_le_bytes()
            )
            .void_unwrap();
        }
    }
```

<details><summary>Output:</summary>
<pre>
[5, 0, 0, 0] / [1, 0, 0, 0] = [1, 0, 1, 0]
[5, 0, 0, 0] / [2, 0, 0, 0] = [1, 0, 1, 0]
[5, 0, 0, 0] / [3, 0, 0, 0] = [1, 0, 1, 0]
[5, 0, 0, 0] / [4, 0, 0, 0] = [1, 0, 1, 0]
[10, 0, 0, 0] / [1, 0, 0, 0] = [1, 0, 1, 0]
[10, 0, 0, 0] / [2, 0, 0, 0] = [1, 0, 1, 0]
[10, 0, 0, 0] / [3, 0, 0, 0] = [1, 0, 1, 0]
[10, 0, 0, 0] / [4, 0, 0, 0] = [1, 0, 1, 0]
[15, 0, 0, 0] / [1, 0, 0, 0] = [1, 0, 1, 0]
[15, 0, 0, 0] / [2, 0, 0, 0] = [1, 0, 1, 0]
[15, 0, 0, 0] / [3, 0, 0, 0] = [1, 0, 1, 0]
[15, 0, 0, 0] / [4, 0, 0, 0] = [1, 0, 1, 0]
[20, 0, 0, 0] / [1, 0, 0, 0] = [1, 0, 1, 0]
...
[80, 0, 0, 0] / [4, 0, 0, 0] = [1, 0, 1, 0]
[85, 0, 0, 0] / [1, 0, 0, 0] = [1, 0, 1, 0]
[85, 0, 0, 0] / [2, 0, 0, 0] = [1, 0, 1, 0]
[85, 0, 0, 0] / [3, 0, 0, 0] = [1, 0, 1, 0]
[85, 0, 0, 0] / [4, 0, 0, 0] = [1, 0, 1, 0]
[90, 0, 0, 0] / [1, 0, 0, 0] = [1, 0, 1, 0]
[90, 0, 0, 0] / [2, 0, 0, 0] = [1, 0, 1, 0]
[90, 0, 0, 0] / [3, 0, 0, 0] = [1, 0, 1, 0]
[90, 0, 0, 0] / [4, 0, 0, 0] = [1, 0, 1, 0]
[95, 0, 0, 0] / [1, 0, 0, 0] = [1, 0, 1, 0]
[95, 0, 0, 0] / [2, 0, 0, 0] = [1, 0, 1, 0]
[95, 0, 0, 0] / [3, 0, 0, 0] = [1, 0, 1, 0]
[95, 0, 0, 0] / [4, 0, 0, 0] = [1, 0, 1, 0]
</pre>
</details>

However:

```rust
    for x in (5u16..100u16).step_by(5) {
        for y in 1u16..5u16 {
            let z: u32 = x as u32 / y as u32;
            uwriteln!(&mut serial, "{} / {} = {:?}\r", x, y, z.to_le_bytes()).void_unwrap();
        }
    }
```

<details><summary>Output:</summary>
<pre>
5 / 1 = [5, 0, 0, 0]
5 / 2 = [2, 0, 0, 0]
5 / 3 = [1, 0, 0, 0]
5 / 4 = [1, 0, 0, 0]
10 / 1 = [10, 0, 0, 0]
10 / 2 = [5, 0, 0, 0]
10 / 3 = [3, 0, 0, 0]
10 / 4 = [2, 0, 0, 0]
15 / 1 = [15, 0, 0, 0]
15 / 2 = [7, 0, 0, 0]
15 / 3 = [5, 0, 0, 0]
15 / 4 = [3, 0, 0, 0]
20 / 1 = [20, 0, 0, 0]
20 / 2 = [10, 0, 0, 0]
20 / 3 = [6, 0, 0, 0]
20 / 4 = [5, 0, 0, 0]
25 / 1 = [25, 0, 0, 0]
25 / 2 = [12, 0, 0, 0]
25 / 3 = [8, 0, 0, 0]
25 / 4 = [6, 0, 0, 0]
30 / 1 = [30, 0, 0, 0]
30 / 2 = [15, 0, 0, 0]
30 / 3 = [10, 0, 0, 0]
30 / 4 = [7, 0, 0, 0]
35 / 1 = [35, 0, 0, 0]
35 / 2 = [17, 0, 0, 0]
35 / 3 = [11, 0, 0, 0]
35 / 4 = [8, 0, 0, 0]
40 / 1 = [40, 0, 0, 0]
40 / 2 = [20, 0, 0, 0]
40 / 3 = [13, 0, 0, 0]
40 / 4 = [10, 0, 0, 0]
45 / 1 = [45, 0, 0, 0]
45 / 2 = [22, 0, 0, 0]
45 / 3 = [15, 0, 0, 0]
45 / 4 = [11, 0, 0, 0]
50 / 1 = [50, 0, 0, 0]
50 / 2 = [25, 0, 0, 0]
50 / 3 = [16, 0, 0, 0]
50 / 4 = [12, 0, 0, 0]
55 / 1 = [55, 0, 0, 0]
55 / 2 = [27, 0, 0, 0]
55 / 3 = [18, 0, 0, 0]
55 / 4 = [13, 0, 0, 0]
60 / 1 = [60, 0, 0, 0]
60 / 2 = [30, 0, 0, 0]
60 / 3 = [20, 0, 0, 0]
60 / 4 = [15, 0, 0, 0]
65 / 1 = [65, 0, 0, 0]
65 / 2 = [32, 0, 0, 0]
65 / 3 = [21, 0, 0, 0]
65 / 4 = [16, 0, 0, 0]
70 / 1 = [70, 0, 0, 0]
70 / 2 = [35, 0, 0, 0]
70 / 3 = [23, 0, 0, 0]
70 / 4 = [17, 0, 0, 0]
75 / 1 = [75, 0, 0, 0]
75 / 2 = [37, 0, 0, 0]
75 / 3 = [25, 0, 0, 0]
75 / 4 = [18, 0, 0, 0]
80 / 1 = [80, 0, 0, 0]
80 / 2 = [40, 0, 0, 0]
80 / 3 = [26, 0, 0, 0]
80 / 4 = [20, 0, 0, 0]
85 / 1 = [85, 0, 0, 0]
85 / 2 = [42, 0, 0, 0]
85 / 3 = [28, 0, 0, 0]
85 / 4 = [21, 0, 0, 0]
90 / 1 = [90, 0, 0, 0]
90 / 2 = [45, 0, 0, 0]
90 / 3 = [30, 0, 0, 0]
90 / 4 = [22, 0, 0, 0]
95 / 1 = [95, 0, 0, 0]
95 / 2 = [47, 0, 0, 0]
95 / 3 = [31, 0, 0, 0]
95 / 4 = [23, 0, 0, 0]
</pre>
</details>

**Comment by kolbma** [comment]
Yes, it is a problem of handling >16bit data for the architecture.  
The instructions need 8-bit or some support 16-bit, the rest need to be handled by code from the compiler.  
So I don't know the format of this data-layout-spec, I couldn't find much info about it some time ago, but the only possibility is to use 8bit "fields" for data to handle bigger ones.  

I've really strange results here when using u32 data just in a calculation and doing some pin-pong with the bits on output pins.
Next to this a LCD is displaying strange output. But the display code is completely __unrelated__ to the u32-calculation. I also use safe-code myself. So the only possibility looks like the wrong memory addresses gets overwritten by the u32-stuff.  
The irony is, rust wants to prevent this.
This would also explain your observation with the size of numbers. Numbers using the first 16-bits are ok, but onwards the data might be put in outland and is not handled correctly.  
Do you have checked multiplication or addition instructions going through the 16-bit border?  
You could isolate if the size of data is a problem or just the division-code doing something unsafe wrong.

**Comment by kolbma** [comment]
I've replaced my rust-core division with https://docs.rs/num-integer/0.1.44/num_integer/fn.div_floor.html
```
num-integer = { version = "0.1", default-features = false }
use num_integer::{div_floor, Integer};
```

Not sure if this works for all data sizes, but in my case the memory corruption goes away with this more or less drop-in replacement.

**Comment by davystrong** [comment]
Also facing this issue.

**Comment by Patryk27** [comment]
Seems like it's due to LTO -- at least in my case using:
```toml
# Cargo.toml

[profile.release]
panic = "abort"
codegen-units = 1
opt-level = "z"
lto = false
```
... fixes the issue for u32, i32, u64 & i64 (128-bit numbers still seem to crash tho').

**Comment by Patryk27** [comment]
This seems to be related to https://github.com/rust-lang/rust/issues/77541 (or at least both seem to be caused by the same underlying bug).

A somewhat simplified example:

```rust
#![no_std]
#![no_main]
#![feature(bench_black_box)]

use atmega_hal::{pins, usart::Usart0, Peripherals};
use core::hint::black_box;
use panic_halt as _;
use ufmt::*;
use void::{ResultVoidExt, Void};

pub type Clock = atmega_hal::clock::MHz20;

#[atmega_hal::entry]
fn main() -> ! {
    let dp = Peripherals::take().unwrap();
    let pins = pins!(dp);

    let mut serial =
        Usart0::<Clock>::new(dp.USART0, pins.pd0, pins.pd1.into_output(), 115200.into());

    uwriteln!(serial, "a").void_unwrap();

    fun(&mut serial, 23, 23);
    fun(&mut serial, 24, 24);
    fun(&mut serial, 25, 25); // if you comment this line, the code works

    uwriteln!(serial, "b").void_unwrap();

    loop {}
}

#[inline(never)]
fn fun(serial: &mut dyn uWrite<Error = Void>, x: u32, y: u32) {
    let _ = black_box(serial);
    let _ = black_box(x / y);
}
```
```toml
[dependencies]
atmega-hal = { git = "https://github.com/Rahix/avr-hal", features = ["atmega328p", "rt"] }
embedded-hal = "0.2"
panic-halt = "0.2"
ufmt = "0.1"
void = { version = "1", default-features = false }

[profile.release]
panic = "abort"
codegen-units = 1
opt-level = 1
lto = false
```
... which you can run using `simavr`:
```
cargo build --release
simavr -m atmega328p -f 20000000 -v target/avr-atmega328p/release/app.elf
```
(requires patched LLVM - https://reviews.llvm.org/D114611)

I think what we're dealing with is a miscompilation that ultimately leads to a stack corruption - investigating `__udivmodsi4` during `25 / 25` shows that the function starts with a correct frame:

![image](https://user-images.githubusercontent.com/3395477/146550808-ede1f45e-90d8-4b8b-bf4c-caa93e9b16f6.png)

... but near the end it does `std` on uninitialized `Z`:

![image](https://user-images.githubusercontent.com/3395477/146550899-33210800-805f-44bc-90a5-47ba20191dac.png)

... which corrupts the frame:

![image](https://user-images.githubusercontent.com/3395477/146551052-600bf62b-c971-419a-a14a-a9dc8da1db4d.png)

Possible cause: https://reviews.llvm.org/D114611#3178558.

**Comment by kolbma** [comment]
@Patryk27 Where is your latest comment I got sent by GH email with the links to  https://reviews.llvm.org/D114611#3178558?

Just some not bug-related question... what might be a good starting point to understand/learn how Rust -> LLVM -> Machine code works?  
How would you start to debug such bugs and find the cause... 
Is there a Must-Read-Book or kind-of I should start with or are there some nice introduction links?
Thx.

**Comment by Patryk27** [comment]
@kolbma I wrote a comment about how with the https://reviews.llvm.org/D114611 patch applied, the AVR backend miscompiles `__udivmodsi4`, leading to stack corruption:

![image](https://user-images.githubusercontent.com/3395477/146566194-2b11148d-134f-46c1-8e93-c58b8081c057.png)

(those `st` + `std` refer to uninitialized `Z`, which effectively scrambles the memory and destroys the stack frame; then, when the function finishes working, `ret` jumps back to 0x00 or whatever address happened to be on the stack.)

... but then I realized that it's probably an oversight in that particular patch of mine (as observed by https://reviews.llvm.org/D114611#3178558) rather than a general bug in the backend, so I figured it's better to delete the comment here and focus on that LLVM's merge request 🙂 

> what might be a good starting point to understand/learn how Rust -> LLVM -> Machine code works?

I've just downloaded LLVM and started fiddling around it; still don't understand ~90% of the AVR's codebase, but commenting stuff out and seeing which tests break is a nice way to learn how various pieces of code come together.

> How would you start to debug such bugs and find the cause...

In this particular case I've cross-compiled the Rust code for AVR and launched it under `simavr -g` + `gdb`, which allowed me to see the assembly and investigate how it works; with a debugger at hand, finding out when the stack frame got corrupted was relatively easy.

**Comment by Patryk27** [comment]
I think I got it!

tl;dr: for a workaround, drop this to `.cargo/config`:
```toml
[unstable]
build-std = ["core"]
build-std-features = ["compiler-builtins-mangled-names"]
```
(128-bit operations will fail to link, but 8/16/32 signed/unsigned should work.)

Explanation:

Since AVR doesn't have a native instruction for division, when the LLVM AVR backend sees a division operation:
```llvm
%quot = sdiv i16 %a, %b
```
... it expands it to a call:
```asm
call __divmodsi4
```
This `__divmodsi4` (and a few other, similar functions) is expected to look more or less like this:
```rust
fn __divmodsi4(a: u32, b: u32) -> (u32, u32) {
  /* ... */
}
```
... and it's expected to be linked fom `avr-gcc` runtime library (https://github.com/gcc-mirror/gcc/blob/31048012db98f5ec9c2ba537bfd850374bdd771f/libgcc/config/avr/lib1funcs.S#L1633).

But, what apparently happens, is that Rust's https://github.com/rust-lang/compiler-builtins, who happens to also provide [a very similar function](https://github.com/rust-lang/compiler-builtins/blob/ea0cb5b589cc498d629c545e9bae600301ba6aed/src/int/udiv.rs#L23), takes precedence:
```rust
fn __udivmodsi4(n: u32, d: u32, rem: Option<&mut u32>) -> u32 {
  /* ... */
}
```
So LLVM AVR's backend is generating code that expects to be linked with avr-gcc's `__divmodsi4`, but in reality it's getting the compiler-builtin's one, which has entirely different signature and just happens to be sometimes aligning with the generated code.

Adding `build-std-features = ["compiler-builtins-mangled-names"]` help to prevent (well, avoid...) this issue, because it causes Rust's `__udivmodsi4` to be linked under a different name, making it possible for avr-gcc to put its own `__udivmodsi4` there.
